### ISSUE TITLE
Build a single Mac distribution

### ISSUE THREAD

**Issue by bradallred** [gemrb/gemrb#1863]
Currently, we are building x86 and arm separately. The reason for this is because we are using homebrew for sourcing packages.

However, Homebrew is a poor choice for actual development on Mac. The devs have stated numerous times that they will not support dual architectures as anything other than isolated casks. Homebrew is also for building unix packages, not Mac packages. This has annoying side effects like having to manually manipulate the `rpath` (#1861)

I know we aren't using GitHub runners (cant remember the reason), but they do [offer](https://github.com/actions/runner-images/blob/main/images/macos/macos-13-Readme.md) both CocoaPods and vcpkg, and I believe both of those are more geared towards development and should support both cross compilation and macOS-isms like frameworks.

we also have a signing issue, but thats its own issue and I'm not yet sure the best course to take there. using `xcodebuild` instead of `cmake` would at least sign the build, but I do like having cmake so maybe its worth the pain of adding it there.

I know cmake can be used to generate an Xcode project and that might work for us. Historically, we have our own Xcode because cmake wasn't quite good enough for our complex array of targets among other things. We also used to have high demand for the iOS build which cmake couldn't do either, but since the EEs nobody has asked about it so maybe we shouldn't care about it (and cmake may work now anyway).

To cross compile with cmake, we need to make sure that `CMAKE_OSX_ARCHITECTURES` is set to `arm64;x86_64`

**Comment by lynxlynxlynx** [comment]
The reason for not using a runner is that they don't provide ARM hardware for mac and that brew can't be used to crosscompile.

**Comment by bradallred** [comment]
you dont need ARM hardware to cross compile, either ARM or x86 will produce what we want. The issue is just homebrew.

**Comment by bradallred** [comment]
I dont understand that response, of course you need a computer to build software. I dont think thats the point you were trying to make tho :)

**Comment by tomkidd** [comment]
Hi there, I run the site Mac Source Ports and I've pulled off a Universal build of GemRB, though not without some issues.

https://macsourceports.com/sourceport/gemrb

The version on the site is still 0.9, I need to make a new build for 0.9.2 I'm just behind on things due to real life issues. 

The way I accomplished this was to build two versions of the app bundle, one for each architecture, and then construct a third bundle using `lipo` and a few other techniques. It's something I learned from the ioquake3 project and it's sort of a "pencil in outer space" solution in that it's crude but it works. 

On an Apple Silicon Mac, you can actually install two parallel versions of Homebrew, since Homebrew keeps the Intel and ARM64 binaries in separate locations. They keep Intel stuff in `/usr/local` and Apple Silicon stuff in `/opt/Homebrew`. You can install the two different versions of Homebrew in parallel too and use an alias to switch between the two. I document the process [here](https://github.com/MacSourcePorts/MSPBuildSystem/wiki/setup), mostly to remind myself how to set up a new system.

The issue I ran into was that, while the `gemrb` executable and the `.dylib` files could be lipo'd together, no such concept exists for the `.so` files. Really, the Universal concept only exists in the first place because Apple keeps changing architectures and being the only full stack vendor they came up with the thing to assist with the transitions. So I had to finagle a concept whereby I would put two parallel directories, one for arm64 and one for x86_64, put the platform-specific version of the `.so` files, and then [modify the code](https://github.com/MacSourcePorts/gemrb/commit/21f96be2ee48589f89f5b658830762c7a9864826) to know where to look. I see that the Windows version of GemRB uses `.dll` files, so I wonder if the `.so` files could be changed to `.dylib` files and then the Universal concept could apply there too, but I haven't looked into it further. 

It worked, but even after all of that, plus signing and notarizing the result, I still got a Gatekeeper issue on first run and I'm not sure why but I figured I'd come back to it later, just never have. 

The problem with modifying the code though is that it makes future updates problematic, and I wondered if the hassle was worth it, so when I migrated to a new build script I just went off of the main GitHub repo and just did parallel builds. I'm not sure why those never made it to the site.

Homebrew is the most popular package manager on macOS, but the problem I have with it is that they do a thing where they maintain the libraries for the version you're running on. So if you're on macOS 13, it gives you macOS 13 linked and optimized libraries. Which for the purposes of running things locally is probably ideal, the problem is that if you're trying to cast a wide net on compatibility (like the source port for an old game, running on an old version of macOS is very on-brand), ABI compatibility is periodically broken. You can put your macOS linking version at, say, 10.9 and link with the Homebrew libraries built for macOS 13 and they'll work on macOS 13, and macOS 12, etc. but anything lower than macOS 10.15 doesn't work because there was a breaking change with how .dylibs work and macOS 10.14 and before don't know about it. I'm not sure what the long term solution is here (short of building all your own libraries yourself, which for a single project is feasible but for someone making dozens of builds, not so much) but it's a known issue. 

Anyway let me know if I can be of any help. I put all the code and scripts I do online but they're not always clear or easy to read, but I do have everything from building to signing to notarizing scripted out.

**Comment by bradallred** [comment]
yes, I know about these techniques, but I am aiming for something less irksome :) If we can solve this by simply suing a different package manager than why not? It doesn't stop anybody from using homebrew for their own builds.

> The issue I ran into was that, while the gemrb executable and the .dylib files could be lipo'd together, no such concept exists for the .so files

`.so`s are the exact same thing as `.dylib`. The extension is chosen by cmake. You can override it with `SMAKE_SHARED_LIBRARY_SUFFIX`. I don't know why cmake chooses `.so` on Mac for "MODULE" libraries, but extensions don't really matter.

We need a separate issue for signing/gatekeeper issues. Feel free to open one now if you have some insights, otherwise I'll do it once the build issues are fixed.

**Comment by lynxlynxlynx** [comment]
Btw, cmake does support signing, just not sure if also with the makefile generator. There's also the bundle generator for cpack (which we don't use) that also does signing: https://cmake.org/cmake/help/git-stage/cpack_gen/bundle.html#

vcpkg has good platform support, but even with an easier time generating the two builds, there's still a need for the glue.

Does mac still support static builds?

**Comment by lynxlynxlynx** [comment]
We set a low CMAKE_OSX_DEPLOYMENT_TARGET already and it supports env vars, so it could be applied to our deps — those that build with cmake. VCPKG has `VCPKG_OSX_DEPLOYMENT_TARGET` with the same purpose and [limitation](https://github.com/microsoft/vcpkg/issues/10038), but workarounds exist.

https://vcpkg.io/en/getting-started.html?platform=mac says macOS 10.14 or below also needs Homebrew, but other than that it seems vcpkg should at least theoretically be able to build for specific versions. By using two triplets we could get both builds and retry merging with `lipo`.

**Comment by bradallred** [comment]
> Does mac still support static builds?

it does and thats always a fine option for distribution.

>  There's also the bundle generator for cpack

yes, and we should probably remove the Mac bundle stuff from our cmake build and use cpack to create the bundle.

Yeah, we can use lipo if we have to, I have never had to for other things. Setting `CMAKE_OSX_ARCHITECTURES=arm64;x86_64` has always been enough provided you have the dependancies for both arches.

**Comment by lynxlynxlynx** [comment]
Mac Source Ports now ships GemRB for all the original games, both for arm and intel CPUs. So at least for releases, users have a better option now:
https://www.macsourceports.com/tag/Infinity%20Engine

Thanks Tom! Let us know if the process has changed in any way, if we should apply any patches etc.

**Comment by tomkidd** [comment]
OK so yeah, in the course of [revamping my build process](https://schnapple.com/2024/11/04/the-mac-source-ports-build-server/), I fell back to the position of just building the two different versions of the GemRB project because I was on the impression that the .so files built in the Mac version weren't Universal 2 material but now I see that comment about how it's really building .dylib files so I may be able to make it Universal 2 after all. I'll check on that at some point. 

My new build process doesn't use Homebrew libraries in favor of manually built libraries that (theoretically) work as far back as OS X 10.7 and are Universal 2 so I don't run into the packaging issues others do, so I might be able to make this whole thing go in one build. 

The tricky part was to bundle a minimal version of Python so that people on a Mac without Homebrew or command line experiences can just run it out of the box. I learned a lot about how the hardened runtime is strict with broken symlinks. 

In any event I can hopefully easily keep new builds happening going forward.

**Comment by tomkidd** [comment]
OK, that was easier than I thought - the .so files being really .dylibs in disguise worked as Universal 2 just fine. 

If anyone wants to try it out here's a Universal 2 build (not on my site yet)
https://github.com/MacSourcePorts/MSPBuildSystem/releases/download/GemRB-0.9.3/gemrb-0.9.3.dmg

**Comment by lynxlynxlynx** [comment]
Glad to hear. I don't see the updated scripts yet on https://github.com/MacSourcePorts/MSPBuildSystem/tree/main/gemrb , but it will be interesting to see the difference.

**Comment by tomkidd** [comment]
OK it's up now. I only modified the "buildsystem" path (so it only builds that part if I pass in a "buildserver" parameter), otherwise it's just the ARM64 script again. 

https://github.com/MacSourcePorts/MSPBuildSystem/blob/buildserver/gemrb/macsourceports_universal2.sh

**Comment by lynxlynxlynx** [comment]
I don't see how to make it easier for you in any significatn way. One simple thing would be, if you pass let's say -DMSP=1 to cmake, I can remove the need for those two sed calls.

**Comment by tomkidd** [comment]
Yeah I'll do a PR at some point - the sed calls are fixing what I think is a bug in CMakeLists - it's modifying the install name of the original dylibs, not the ones copied to the app bundle, which is what I think it's meant to be doing.

**Comment by lynxlynxlynx** [comment]
Cool, looking forward to that. And oops, referenced the wrong issue.
